File: elixir/perfect-numbers/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(1), do: {:ok, :deficient}
  def classify(number) when is_integer(number) and number > 1 do
    aliquot = number |> factors() |> Enum.sum()
    cond do
      number == aliquot -> {:ok, :perfect}
      number < aliquot -> {:ok, :abundant}
      number > aliquot -> {:ok, :deficient}
    end
  end
  def classify(_), do: {:error, "Classification is only possible for natural numbers."}

  defp factors(number), do: 1..number - 1 |> Enum.filter(& rem(number, &1) == 0)
end

File: python/perfect-numbers/perfect_numbers.py
def classify(number):
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """
    if number < 1:
        raise ValueError('Classification is only possible for positive integers.')

    aliquot = sum(num for num in range(1, number // 2 + 1) if number % num == 0)

    return 'abundant' if number < aliquot else 'deficient' if number > aliquot else 'perfect'
